Add a link checker to k8s.io (#20606)
kubernetes/website: content/en/docs/doc-contributor-tools/linkchecker/README.md

# Internal link checking tool

> You can use the https://github.com/wjdp/htmltest tool to check for broken links on **english content**. This is a helpful when refactoring sections of content, moving pages around, or renaming files or page headers.

Reviewer: Omit needless words, capitalize, "helpful" is anthropomorphic
```suggestion
You can use [htmltest](https://github.com/wjdp/htmltest) to check for broken links on **English content**. This is useful when refactoring sections of content, moving pages around, or renaming files or page headers.
```

> ## What is and isn't checked

Reviewer: Eliminate passive voice
```suggestion
## What it does and doesn't check
```

> The link checker scans generated HTML files, **not** raw markdown. You can see the full configuration in the `.htmltest.yml` file in the root of `kubernetes/website`.

Reviewer: Remove emphasis, add link
```suggestion
The link checker scans generated HTML files, not raw markdown. You can see the full configuration in [`.htmltest.yml`](https://git.k8s.io/website/.htmltest.yml).
```

The link checker scans the following:

> - All content generated from markdown in the `/content/en/docs` directory **excluding**:

Reviewer: Capitalize, link, omit needless words, remove emphasis
```suggestion
- All content generated from Markdown in [`/content/en/docs`](https://git.k8s.io/website/content/en/docs/) directory, excluding:
```

Author: TIL that `git.k8s.io` is a thing

  - Generated API references

> - All internal links **excluding**:

Reviewer: Remove emphasis, use commas consistently
```suggestion
- All internal links, excluding:
```

  - Empty hashes (`<a href="#">` or `[title](#)`) and empty hrefs (`<a href="">` or `[title]()`)
  - Internal links to images and other media files

> The link checker **does not** scan the following:

Reviewer: Remove emphasis
```suggestion
The link checker does not scan the following:
```

> - Links included in the site's "chrome": top and side menu links, footer links, links in a page's `<head>` section such as those to CSS stylesheets, scripts, and meta information

Reviewer: Eliminate jargon
```suggestion
- Links included in the top and side nav bars, footer links, or links in a page's `<head>` section, such as links to CSS stylesheets, scripts, and meta information
```

> - Top level pages _and_ their children, for example: `/training`, `/community`, `/case-studies/adidas`

Reviewer: I'm squinting hard at those italics, but I leave it to your discretion.

![buffy](https://user-images.githubusercontent.com/3210446/81775482-c5f47800-94a1-11ea-8bbe-773e50ea7557.gif)

- Blog posts
- API Reference documentation
- Localizations

## Prerequisites and installation

> [Docker](https://docs.docker.com/get-docker/) installed locally.

Reviewer: ```suggestion
You must install [Docker](https://docs.docker.com/get-docker/) locally.
```

> The tool runs using a `make` command, so that's it!

Reviewer: Omit needless words
```suggestion
The tool runs using `make`, so that's it!
```

## Running the link checker

To run the link checker:

> 1. Navigate to the root directory of your local `kubernetes/website` repository.

Reviewer: Good. 👍

2. Run the following command:

  ```
  make check-internal-links
  ```

> ## Understanding the output

Reviewer: Nice heading title! 👍

If the link checker finds broken links, the output is similar to the following:

```
tasks/access-kubernetes-api/custom-resources/index.html
  hash does not exist --- tasks/access-kubernetes-api/custom-resources/index.html --> #preserving-unknown-fields
  hash does not exist --- tasks/access-kubernetes-api/custom-resources/index.html --> #preserving-unknown-fields
```

> This is one "set" of broken links. More output is added to the log for each page with broken links.

Reviewer: No quotes needed, eliminate passive voice
```suggestion
This is one set of broken links. The log adds an output for each page with broken links.
```

> In this output:
> 
> - The file with broken links is `tasks/access-kubernetes-api/custom-resources.md`.
> - The reason is `hash does not exist`. In most cases, you can ignore this.
> - The URL linked to is `#preserving-unknown-fields`.

Reviewer: Only make lists if you're enumerating things.

Handy vocabulary from my SEO past: "source URL" and "target URL" for origin and destination.
```suggestion
In this output, the file with broken links is `tasks/access-kubernetes-api/custom-resources.md`.

The tool gives a reason: `hash does not exist`. In most cases, you can ignore this.

The target URL is `#preserving-unknown-fields`.
```

> The best way to fix this is to:

Reviewer: Avoid value judgments
```suggestion
One way to fix this is to:
```

> 1. Navigate to the markdown file in question.

Reviewer: ```suggestion
1. Navigate to the file with the broken link.
```

Author: :+1:, but I'll probably specify "the markdown file" in some way with this sentence. 

The tool scans over rendered HTML files, and therefore lists all its results as 
`tasks/access-kubernetes-api/custom-resources/index.html`. I think it's important to be crystal clear that we need people to fix the markdown, not the rendered HTML.

Reviewer: Fair 👍, as long as it's "the Markdown file", because the spec is capitalized.

2. Using a text editor, do a full-text search (usually Ctrl+F or Command+F) for the broken link's URL, `#preserving-unknown-fields`.

> 3. Address the problem. For a broken page hash (anchor) link, the first thing to check is the rest of the page's topics to see if a topic has been renamed.

Reviewer: Italics when introducing a new term, use plain language
```suggestion
3. Fix the link. For a broken page hash (or _anchor_) link, check whether the topic was renamed or removed.
```

> 4. Run the link checker again until there are no broken links repots left.

Reviewer: ```suggestion

Run htmltest to verify that broken links are fixed.
```

Reviewer: There's a spelling error in the current version, so this isn't resolved.

# Internal link checking tool

> 

Reviewer 2: nit: Internal == intra-site and only English?

> You can use [htmltest](https://github.com/wjdp/htmltest) to check for broken links in [`/content/en/`](https://git.k8s.io/website/content/en/). This is useful when refactoring sections of content, moving pages around, or renaming files or page headers.

Reviewer 2: Could explain the how the tool works and the limitations (English only?):

[htmltest](https://github.com/wjdp/htmltest) is a tool to check for broken links in HTML output.
Steps to use the tool in this site:
build the HTML content,
pull the `wjdp/htmltest` docker image 
run the tool over the generated HTML content, [`/content/en/`](https://git.k8s.io/website/content/en/). 
Generate a log file (location). Review the log and make corrections?
This is useful when refactoring sections of content, moving pages around, or renaming files or page headers.

## What it does and doesn't check

> The link checker scans generated HTML files, not raw markdown. You can see the full configuration in [`.htmltest.yml`](https://git.k8s.io/website/.htmltest.yml).

Reviewer 2: Where are the generated HTML files located. Do contributors know where to look?
annoying nit: Markdown (be consistent on page)

The htmltest tool depends on a configuration file, [`.htmltest.yml`](...) to determine which content to examine.

The link checker scans the following:

- All content generated from Markdown in [`/content/en/docs`](https://git.k8s.io/website/content/en/docs/) directory, excluding:

>   - Generated API references

Reviewer 2: Line 12, `Generated API references` is referring to the individual component tool reference pages (which are .md files)?

Reviewer 2: What's the difference between scanning `All content generated` and `All internal links`?

Author: @kbhawkey line 12: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/ & older versions of the API reference aren't scanned. (Specifically, anything that lands in the `reference/generated/kubernetes-api` directory in the site's final build is ignored)

- All internal links, excluding:
  - Empty hashes (`<a href="#">` or `[title](#)`) and empty hrefs (`<a href="">` or `[title]()`)
  - Internal links to images and other media files

The link checker does not scan the following:

- Links included in the top and side nav bars, footer links, or links in a page's `<head>` section, such as links to CSS stylesheets, scripts, and meta information
- Top level pages and their children, for example: `/training`, `/community`, `/case-studies/adidas`
- Blog posts

> - API Reference documentation

Reviewer 2: Line 22, meaning the two HTML references (kubectl-command and API ref)?

Author: Once again – https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/ and its ilk. Really just reiterating here to drive the point home.

- Localizations

## Prerequisites and installation

You must install [Docker](https://docs.docker.com/get-docker/) locally.

> 

Reviewer 2: You must install:
- `docker`
- `make`

1. Navigate to the root directory of your local `kubernetes/website` repository.

2. Run the following command:

  ```
  make check-internal-links
  ```

## Understanding the output

If the link checker finds broken links, the output is similar to the following:

```
tasks/access-kubernetes-api/custom-resources/index.html
  hash does not exist --- tasks/access-kubernetes-api/custom-resources/index.html --> #preserving-unknown-fields
  hash does not exist --- tasks/access-kubernetes-api/custom-resources/index.html --> #preserving-unknown-fields
```

This is one set of broken links. The log adds an output for each page with broken links.

In this output, the file with broken links is `tasks/access-kubernetes-api/custom-resources.md`.

The tool gives a reason: `hash does not exist`. In most cases, you can ignore this.

The target URL is `#preserving-unknown-fields`.

One way to fix this is to:

1. Navigate to the Markdown file with a broken link.
2. Using a text editor, do a full-text search (usually Ctrl+F or Command+F) for the broken link's URL, `#preserving-unknown-fields`.
3. Fix the link. For a broken page hash (or _anchor_) link, check whether the topic was renamed or removed.

> Run htmltest to verify that broken links are fixed.

Reviewer 2: nit: Re-run htmltest ?

- Scans the files generated in the `/public` directory and provides command line output when it encounters broken internal links

## What it does and doesn't check

The link checker scans generated HTML files, not raw Markdown. The htmltest tool depends on a configuration file, [`.htmltest.yml`](https://git.k8s.io/website/.htmltest.yml), to determine which content to examine.

The link checker scans the following:

- All content generated from Markdown in [`/content/en/docs`](https://git.k8s.io/website/content/en/docs/) directory, excluding:
  - Generated API references, for example https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/
- All internal links, excluding:
  - Empty hashes (`<a href="#">` or `[title](#)`) and empty hrefs (`<a href="">` or `[title]()`)
  - Internal links to images and other media files

The link checker does not scan the following:

- Links included in the top and side nav bars, footer links, or links in a page's `<head>` section, such as links to CSS stylesheets, scripts, and meta information
- Top level pages and their children, for example: `/training`, `/community`, `/case-studies/adidas`
- Blog posts

> - API Reference documentation, , for example https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/

Reviewer 3: nit:
```suggestion
- API Reference documentation; for example: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/
```

Author: Resolved locally ergo not committing this suggestion :)

- Localizations

## Prerequisites and installation

You must install
* [Docker](https://docs.docker.com/get-docker/)

> * [make](https://www.gnu.org/software/make/)

Reviewer 3: It looks like `make check-internal-links` also requires Hugo.
